pyquil/api/_client.py

```diff
@@ -0,0 +1,241 @@
+               client will be created using ``configuration``.
+        :param configuration: QCS configuration.
+        """
+        self._config = configuration or QCSClientConfiguration.load()
+        self._has_default_config = configuration is None
+        self._http = http
+
+    def post_json(self, url: str, json: Any, timeout: float = 5) -> httpx.Response:
+        """
+        Post JSON to a URL. Will raise an exception for response statuses >= 400.
+
+        :param url: URL to post to.
+        :param json: JSON body of request.
+        :param timeout: Time limit for request, in seconds.
+        :return: HTTP response corresponding to request.
+        """
+        logger.debug("Sending POST request to %s. Body: %s", url, json)
+        with self._http_client() as http:  # type: httpx.Client
+            res = http.post(url, json=json, timeout=timeout)
+            if res.status_code >= 400:
+                raise _parse_error(res)
+        return res
+
+    def qcs_request(self, request_fn: Callable[..., Response[Any]], **kwargs: Any) -> Any:
+        """
+        Execute a QCS request.
+
+        :param request_fn: Request function (from ``qcs_api_client.operations.sync``).
+        :param kwargs: Arguments to pass to request function.
+        :return: HTTP response corresponding to request.
+        """
+        with self._http_client() as http:  # type: httpx.Client
+            return request_fn(client=http, **kwargs).parsed
+
+    def compiler_rpcq_request(
+        self, method_name: str, *args: Any, timeout: Optional[float] = None, **kwargs: Any
+    ) -> Any:
+        """
+        Execute a remote function against the Quil compiler.
+
+        :param method_name: Method name.
+        :param args: Arguments that will be passed to the remote function.
+        :param timeout: Optional time limit for request, in seconds.
+        :param kwargs: Keyword arguments that will be passed to the remote function.
+        :return: Result from remote function.
+        """
+        return self._rpcq_request(self.quilc_url, method_name, *args, timeout=timeout, **kwargs)
+
+    def processor_rpcq_request(
+        self,
+        processor_id: str,
+        method_name: str,
+        *args: Any,
+        timeout: Optional[float] = None,
+        **kwargs: Any,
+    ) -> Any:
+        """
+        Execute a remote function against a processor endpoint. If there is no current engagement,
+        or if it is invalid, a new engagement will be requested for the given processor.
+
+        :param processor_id: Processor to engage.
+        :param method_name: Method name.
+        :param args: Arguments that will be passed to the remote function.
+        :param timeout: Optional time limit for request, in seconds.
+        :param kwargs: Keyword arguments that will be passed to the remote function.
+        :return: Result from remote function.
+        """
+        # TODO(andrew): handle multiple engagements at once (per processor)
```

Comment:
@kalzoo I'll likely defer this to a later PR after all (when we do more refactoring on the client/engagements)